# **Recreating Figure 3A**
---
## **Inputs:**

| No. | File(s)             | Path                                                                        |Generated by| Used by | Description |
|-----|---------------------|-----------------------------------------------------------------------------|------------|---------|-------------|
|**1.**|{SAMPLE}_CONC.addgp.rmdup.bam` |`pipeline/data/bams/`|A combination between `samtools` and `picard` |`samtools` inside `run.sh`,`overlapping_genesAbove.py`, `ortholog_coverageAll.py`|Genome alignment to a concatenated reference. The index is also necessary.|
|**2.**|`SACE_S288C_v1_allChr.gff`|`../../figure1/fig1C/pipeline/data/ref/`|Obtained from: [Liti, et.al.](https://yjx1217.github.io/Yeast_PacBio_2016/data/)|`gff2txt.py`|Genome anotation from the species to which the samples belong.|
|**3.**|`SAPA_YPS138_v1.gff`|`../../figure1/fig1C/pipeline/data/ref/`|Obtained from: [Liti, et.al.](https://yjx1217.github.io/Yeast_PacBio_2016/data/)|`gff2txt.py`|Genome anotation from the species to which the introgressions come from.|
|**4.**|`SACE_S288C_v1_allChr.fna`|`pipeline/data/`|---|`ortholog_coverageAll.py`|Fasta of the species to wich samples belong.|
|**5.**|`SAPA_YPS138_v1.fna`|`pipeline/data/`|---|`ortholog_coverageAll.py`|Fasta of the species to wich the introgressions belong.|
|**6.**|`SAPA_YPS138_v1_subtelomeric_regions.csv`|`pipeline/data/`|Subtelomeric regions obtained by |`ortholog_coverageAll.py`|Coordinates of the S. paradoxus subtelomeric regions.|
|**7.**|`41588_2017_BFng3847_MOESM339_ESM.tsv`|`pipeline/data/`|One to one ortholos obtained by |`ortholog_coverageAll.py`|A list of one to one orthologs between Saccharomyces species.|
|**8.**|`phylogeny.csv`|`pipeline/`|Manually curated. |`graph.R`|IDs of the strains used ant their phylogenetic groups.|


*Note:* All paths are relative to this file, exceptions are signaled by a starting `/`.*

---
## **Scripts:**
| No. | Script             | Path                                                                        | Output | Description |
|-----|--------------------|-----------------------------------------------------------------------------|--------|-------------|
|**1.**|`run.sh`|`pipeline/`|Multiple sges as well as several bash files on `bin/sge/`|Runs all the pipeline: 1. Compute coverage depth, 2. obtain coverage statistics, 3. filters and converts gff into txt, 4. find genes with coverage and 5. obtain introgressions.|
|**2.**|`make_depth_plot.R`|`bin/scripts/`|CSVs with the coverage statistics per windows, per reference and per contig.|Takes the depth file of the alignment to the concatenated reference and computes coverage depth statistics at different levels.|
|**3.**|`gff2txt.py`|`bin/scripts/`|<p>TXT formatted gff containing only genes.</p><p>Bed file with all genes described in the txt mentioned before.</p><p>Index of the bed file mentioned before in tabix format.</p>|Parses gff files and converts them into files more convenient for this pipeline.|
|**4.**|`overlapping_genesAbove.py`|`bin/scripts/`|CSVs with coverage from genes.|For a specific sample, obtains well covered genes from a given specie.|
|**5.**|`ortholog_coverageAll.py`|`bin/scripts/`|CSVs with coverage statistics from introgressions.|Takes well covered genes from the specie to which introgressions belong and clasifies them as introgressions depending on various statistics.|
|**6.**|`graph.R`|`bin/scripts/`|Graph with the distributions of number of introgressions by phylogenetic group, colored by heterozigosity.|Takes the csvs with the introgressions and graphs figure 3A.|


### *Packages/Dependencies:*
**1.** htslib 1.9

**2.** gcc 5.1.0

**3.** bedtools 2.26

**4.** python 3.7

**5.** samtools 1.9

**6.** htslib 1.2

**7.** r 3.6.1

**8.** bedtools 2.24

---
## **Outputs:**
| No. | File | Path | Description |
|-----|------|------|-------------|
|**1.**|`{SAMPLE}_introSAPA_All_noOrth.csv`|`pipeline/data/introgressions/`|CSV file with introgressions per sample, includes information about the introgressed gene.|
|**2.**|`graph.pdf`|`pipeline/data/introgressions/`|Figure 3A.|


*Note:* Several intermediate files are created but not all of them are stored. Only relevant files are listed above.

---
## **Execution:**
**1.**
```
cd pipeline
bash run.sh
```
