# **Recreating Figure 4A**
---
## **Inputs:**

| No. | File(s)             | Path                                                                        |Generated by| Used by | Description |
|-----|---------------------|-----------------------------------------------------------------------------|------------|---------|-------------|
|**1.**| {SAMPLE}_CONC.g.vcf |VCF of each sample after aligning to the concatenated reference. | --- | `pipeline/Snakefile`| VCFs obatained after aligning the samples fastqs to the concatenated reference.|

*Note:* All paths are relative to this file, exceptions are signaled by a starting `/`.

---
## **Scripts:**
| No. | Script             | Path                                                                        | Output | Description |
|-----|--------------------|-----------------------------------------------------------------------------|--------|-------------|
|**1.**|`run.sh`|`pipeline/`|---|Runs all the pipeline: 1. Creates symbolic links to  all VCFs, 2. Creates introgression blocks for each sample 3. Executes Snakemake pipeline to make the vcfs with biallelic snps.|
|**2.**|`Snakefile`|`pipeline/`|`data/vcfs/{intro}/Matrix_{intro}_combinegvcfs_genotype_onlySNPs_filterlow_filteredSNPs_missing40_biallelic.vcf`|Snakemake pipeline that creates vcfs in blocks of introgression|
|**3.**|`coords.smk`|`pipeline/scripts/`|`data/introgression_coordinates.txt` y `data/shared_introgressions.txt`|Finds coordinates of introgression blocks a sees how they are shared.|
|**4.**|`MakeBlocks.smk`|`pipeline/scripts/`|`data/blocks/{sample}_blocks.csv` y `data/summary_statistics_blocks.csv`|Finds blocks of introgression in samples and obtains summary statistics of each block.|

### *Packages/Dependencies:*
**1.** Python 3.7

**2.** Snakemake 5.10.0

**3.** vcftools 0.1.14

**4.** gatk 4.1.1.0

**5.** bcftools 1.9


---
## **Outputs:**
| No. | File | Path | Description |
|-----|------|------|-------------|
|**1.**|`data/vcfs/{intro}/Matrix_{intro}_combinegvcfs_genotype_onlySNPs_filterlow_filteredSNPs_missing40_biallelic.vcf`|`pipeline/data/vcfs`|Combined VCF with al samples that share an introgression block|

*Note:* Several intermediate files are created but not all of them are stored. Only relevant files are listed above.

---
## **Execution:**
**1.** On the terminal run:
```
cd pipeline
bash run.sh
```
**2.** Visualization was done with [microreact](https://microreact.org).

## **Calculate Paired Distances:**

| No. | File(s)             | Path                                                                        |Generated by| Used by | Description |
|-----|---------------------|-----------------------------------------------------------------------------|------------|---------|-------------|
|**1.**| {data/vcfs/{intro}/Matrix_{intro}_combinegvcfs_genotype_onlySNPs_filterlow_filteredSNPs_missing40_biallelic.vcf	 | run.sh | `Get_Paired_Distances.sh`| Get the number of differences between all pairs of strains in each introgressed-block specific VCF.|

*Note:* All paths are relative to this file, exceptions are signaled by a starting `/`.

### *Packages/Dependencies:*

**1.** bcftools 1.9

---
## **Outputs:**
| No. | File | Path | Description |
|-----|------|------|-------------|
|**1.**|`data/PairedDistances/paired_distances_{intro}.csv`|`pipeline/data/PairedDistances`|List of all pairs of strains in a vcf and the number of differences between them|


---
## **Execution:**
**1.** On the terminal run:
```
cd pipeline
bash Get_Paired_Distances.sh
```
