# **Recreating Figure 1C**
---
## **Inputs:**

| No. | File(s)             | Path                                                                        |Generated by| Used by | Description |
|-----|---------------------|-----------------------------------------------------------------------------|------------|---------|-------------|
|**1.**| {SAMPLE}_CONC.g.vcf |VCF obtained after aligning each samples against a concatenated reference| --- | `pipeline/Snakefile`| VCFs obatained after aligning the samples fastqs to the concatenated reference.|
|**2.**|`SACE_samples.csv`|`pipeline/`|Manually curated|---|<p>CSV with all 467 samples analyzed on the paper.</p><p>Used to generate `SACE467_ToFig3A_mod.csv`</p>|
|**3.**|`SACE467_ToFig3A_mod.csv`|`pipeline/`|Manually curated|`pipeline/run.sh`|<p>CSV with a subet of samples from `SACE_samples.csv`.</p>Admixed populations were manually excluded from this file.</p>|
|**4.**|`list_ortho.txt`|`pipeline/`|Obtained from: [Fischer, et.al.](https://www.nature.com/articles/s41588-023-01459-y)|`pipeline/scripts/make_bed.py`|A list of ortholog groups.|
|**5.**|`SACE_S288C_v1_allChr.gff`|`pipeline/data/ref/`|Obtained from: [Liti, et.al.](https://yjx1217.github.io/Yeast_PacBio_2016/data/)|`pipeline/scripts/make_bed.py`|Genome anotation from the species to which the samples belong.|
|**6.**|`CONC_Mez1_v1_allChr.fasta`|`pipeline/data/ref/`|Manually concatenated genomes|`pipeline/Snakefile`|Concatenated reference.|

*Note:* All paths are relative to this file, exceptions are signaled by a starting `/`.

---
## **Scripts:**
| No. | Script             | Path                                                                        | Output | Description |
|-----|--------------------|-----------------------------------------------------------------------------|--------|-------------|
|**1.**|`run.sh`|`pipeline/`|---|Runs all the pipeline: 1. Creates symbolic links to  all VCFs, 2. Creates bed file from list of orthologs and 3. Executes Snakemake pipeline to make the phylogenetic tree.|
|**2.**|`Snakefile`|`pipeline/`|`data/phyl/RAxML_bestTree.ortho.tree`|Snakemake pipeline that create phylogenetic trees from VCFs.|
|**3.**|`make_bed.py`|`pipeline/scripts/`|`pipeline/data/ortholog_regions.bed`|Takes as input a list of genes and generates a bed file with the coordinates of those genes.|
|**4.**|`vcf2phylip.py`|`pipeline/scripts/`|`pipeline/data/phyl/temp.min4.phy.|Takes VCF file and converts it into phylip format.|

### *Packages/Dependencies:*
**1.** Python 3.7

**2.** Snakemake 5.10.0

**3.** vcftools 0.1.14

**4.** gatk 4.1.1.0

**5.** bcftools 1.9

**6.** raxml 8.2.12-avx2-pthreads

---
## **Outputs:**
| No. | File | Path | Description |
|-----|------|------|-------------|
|**1.**|`RAxML_bestTree.ortho.tree`|`pipeline/data/phyl/`|Phylogenetic tree.|

*Note:* Several intermediate files are created but not all of them are stored. Only relevant files are listed above.

---
## **Execution:**
**1.** On the terminal run:
```
cd pipeline
bash run.sh
```
**2.** Visualization was done with [microreact](https://microreact.org).
